common/k8s/src/event_source.rs

```diff
@@ -0,0 +1,212 @@
+}
+
+impl From<Event> for EventLogLine {
+    // Replicate the Reporter's formatting
+    fn from(event: Event) -> Self {
+        let Event {
+            type_,
+            action,
+            reason,
+            message,
+            count,
+            source,
+            involved_object,
+            last_timestamp,
+            first_timestamp,
+            ..
+        } = event;
+        let ObjectReference {
+            kind,
+            name,
+            namespace,
+            field_path,
+            ..
+        } = involved_object;
+        let (node, component) = if let Some(s) = source {
+            (s.host, s.component)
+        } else {
+            (None, None)
+        };
+
+        let (host, app) = if let Some(kind) = kind.as_ref() {
+            if kind == "Pod" {
+                (
+                    name.clone(),
+                    field_path
+                        .map(|field_path| field_path.split("[{}]").nth(1).map(|s| s.to_string()))
+                        .flatten(),
+                )
+            } else if let Some(ref name) = name {
+                (Some(format!("{}/{}", kind, name)), None)
+            } else {
+                (None, None)
+            }
+        } else {
+            (None, None)
+        };
+        let app = app.or_else(|| component.clone());
+
+        let age = match (last_timestamp.as_ref(), first_timestamp.as_ref()) {
+            (Some(last), Some(first)) if last > first => Some(last.0 - first.0),
+            _ => None,
+        };
+
+        let duration = age.map(|age| {
+            if age > Duration::weeks(0) {
+                HumanTime::from(age).to_string()
+            } else {
+                "just now".to_string()
+            }
+        });
+
+        EventLogLine {
+            level: type_,
+            message: match (
+                reason.as_ref(),
+                count.as_ref(),
+                duration.as_ref(),
+                message.as_ref(),
+            ) {
+                (Some(r), Some(c), Some(d), Some(m)) => {
+                    Some(format!("{}  (x{} {})  {}", r, c, d, m))
+                }
+                _ => None,
+            },
+            _name: host,
+            _app: app,
+            kube: EventLogLineInner {
+                type_: "event".to_string(),
+                action,
+                resource: kind,
+                name,
+                namespace,
+                reason,
+                message,
+                component,
+                node,
+                first_time: first_timestamp,
+                time: last_timestamp,
+                age: age
+                    .map(|age| {
+                        age.num_seconds()
+                            .try_into()
+                            .map_err(|_| warn!("age too large, could not fit {} into i32", age))
+                            .ok()
+                    })
+                    .flatten(),
+                count,
+            },
+        }
```

Comment:
This structure looks good to me. Would be useful if we could compare what this line looks like in the web UI and compare it with the kube reporter - want to make sure everything matches up.

Reply:
Yeh, this is based entirely on the reporter's class/EventWatcher.js handling. I was going to add a link in a comment but it's not in a public repo. I'll add a full example in the unit tests and include the reporter's example payload and test case